lib/phoenix/verified_routes.ex

```diff
@@ -0,0 +1,757 @@
+            "or a Phoenix.Endpoint when building url at #{path}, got: #{inspect(other)}"
+  end
+
+  defp concat_url(url, path) when is_binary(path), do: url <> path
+
+  @doc """
+  Generates path to a static asset given its file path.
+  """
+  def static_path(%Plug.Conn{private: private}, path) do
+    case private do
+      %{phoenix_static_url: _} -> path
+      %{phoenix_endpoint: endpoint} -> endpoint.static_path(path)
+    end
+  end
+
+  def static_path(%URI{} = uri, path) do
+    (uri.path || "") <> path
+  end
+
+  def static_path(%_{endpoint: endpoint}, path) do
+    static_path(endpoint, path)
+  end
+
+  def static_path(endpoint, path) when is_atom(endpoint) do
+    endpoint.static_path(path)
+  end
+
+  @doc """
+  Returns the path with relevant script name prefixes without verification.
+  """
+  def unverified_path(%Plug.Conn{} = conn, router, path) do
+    conn
+    |> build_own_forward_path(router, path)
+    |> Kernel.||(build_conn_forward_path(conn, router, path))
+    |> Kernel.||(path_with_script(path, conn.script_name))
+  end
+
+  def unverified_path(%URI{} = uri, _router, path) do
+    (uri.path || "") <> path
+  end
+
+  def unverified_path(%_{endpoint: endpoint}, router, path) do
+    unverified_path(endpoint, router, path)
+  end
+
+  def unverified_path(endpoint, _router, path) when is_atom(endpoint) do
+    endpoint.path(path)
+  end
+
+  def unverified_path(other, router, path) do
+    raise ArgumentError,
+          "expected a %Plug.Conn{}, a %Phoenix.Socket{}, a %URI{}, a struct with an :endpoint key, " <>
+            "or a Phoenix.Endpoint when building path for #{inspect(router)} at #{path}, got: #{inspect(other)}"
+  end
+
+  @doc false
+  def __encode_segment__(data) do
+    case data do
+      [] -> ""
+      [str | _] when is_binary(str) -> Enum.map_join(data, "/", &encode_segment/1)
+      _ -> encode_segment(data)
+    end
+  end
+
+  defp encode_segment(data) do
+    data
+    |> Phoenix.Param.to_param()
+    |> URI.encode(&URI.char_unreserved?/1)
+  end
+
+  defp verify_segment(["/" | rest], route, acc), do: verify_segment(rest, route, ["/" | acc])
+
+  # we've found a static segment, return to caller with rewritten query if found
+  defp verify_segment(["/" <> _ = segment | rest], route, acc) do
+    case {String.split(segment, "?"), rest} do
+      {[segment], _} ->
+        verify_segment(rest, route, [URI.encode(segment) | acc])
+
+      {[segment, static_query], dynamic_query} ->
+        {Enum.reverse([URI.encode(segment) | acc]),
+         verify_query(dynamic_query, route, [static_query, "?"])}
+    end
+  end
+
+  # we reached the static query string, return to caller
+  defp verify_segment(["?" <> _ = query], _route, acc) do
+    {Enum.reverse(acc), [query]}
+  end
+
+  # we reached the dynamic query string, return to call with rewritten query
+  defp verify_segment(["?" | rest], route, acc) do
+    {Enum.reverse(acc), verify_query(rest, route, ["?"])}
+  end
```

Comment:
I think this case is broken:

    /posts/#{post.id}?foo=bar&#{key}=#{value}&baz=bat

If so, this might be a fix:

```suggestion
  # we reached the dynamic query string, return to call with rewritten query
  defp verify_segment(["?" <> _ = query | rest], route, acc) do
    {Enum.reverse(acc), verify_query(rest, route, [query])}
  end
```